### PR TITLE
fix subsetting for dagster-dbt cloud integration

### LINKED ISSUE #21593: Sub-subsetting does not work for dbt-cloud assets with dbt version after 1.5.0
### Dagster version

dagster, version 1.7.1

### What's the issue?

**TLDR:** The multiple `--select` [hack](https://github.com/dagster-io/dagster/commit/1bbe26e3cf29a5f3f55aadd58a6c4fa8562c037e#diff-ba3879628c3facbf5cdf9130e7320f6b2157b9b60e13091df16cfd762974dbf4R332) used in launching dbt Cloud jobs with a `--select` statement predefined in the job does not work since dbt-core version 1.5.0, which [allows multiple](https://github.com/dbt-labs/dbt-core/releases/tag/v1.5.0) `--select` statements.

**Long version:**
We already have a massive dbt project and are now trying to shift from dbt Cloud orchestration to Dagster orchestration for both pre-DWH and dbt-based steps of our pipelines. So, we are implementing a single pipeline where both parts are orchestrated in Dagster. The pipeline looks pretty simple:

- Asset to load data into GCS from the outer world
- Asset for "source" table in BQ created from GCS data
- Run dbt models rooting from the source table
Since we don't (yet) want to upload the whole dbt project as assets into dbt, we defined an "integration" dbt Cloud job as `dbt build --select source:raw_data.raw_data_table_from_gcs+`.

It worked great when creating assets; we now have all related models as them. However, issue appeared when I tried to materialize only one asset from the middle of the dbt graph. The command sent by Dagster in the dbt Jobs UI in this case is something like `dbt build --select source:raw_data.raw_data_table_from_gcs+ --select asset_i_wanted_to_materialize`. In the dbt log, I can clearly see building all the models from the initial `source:raw_data.raw_data_table_from_gcs+` part of the command, not only the selected model from Dagster.

So, I delved into the dagster_dbt source code to understand how Dagster triggers a job in such cases. What I found is a [part of the method with a comment](https://github.com/dagster-io/dagster/blob/bc81704666099720e8dacc927210301fc0542d22/python_modules/libraries/dagster-dbt/dagster_dbt/cloud/asset_defs.py#L503) stating that dbt uses only the last `--select` argument for run/build commands, so just adding the selection to the command from the job should be fine.

I tested this statement on our current dbt version 1.7 with `dbt ls` and it resulted in a union of both `--select` args instead of only the last one. So my first assumption was that the dagster code for this was written quite a while ago, so I tested it on like version 1.3.1 of dbt, which was state-of-the-art at that moment, and it worked as Dagster expects, returning only `asset_i_wanted_to_materialize`.

Thus, the issue is definitely somewhere in dbt's changes between versions. I spent a couple more minutes and was able to locate the version and [release note](https://github.com/dbt-labs/dbt-core/releases/tag/v1.5.0) exactly about this change in the dbt version 1.5.0. There dbt stopped ignoring non-last --select arguments and (I assume) now unions them all together as just regular multiple --select keys. This change obviously breaks the Dagster assumption and results in a full job run even when only one asset from the job was selected to materialize.

### What did you expect to happen?

I assume more correct way to handle sub-selections for jobs with already pre-defined `--select` argument is using [intersection operator](https://docs.getdbt.com/reference/node-selection/set-operators#intersections).

In our simple case it will work pretty well with just appending `,asset_i_wanted_to_materialize` to initial job selection.
It will also work in case of multiple `--select` statement by adding intersection operator to every argument like `--select source:foo.bar+ source:foo.buz` -->  `--select source:foo.bar+,asset_i_wanted_to_materialize source:foo.buz,asset_i_wanted_to_materialize`.

However, I cannot find a better way to deal with this issue in more complex case, like already used intersection operator, `--exclude` flag etc. It looks like dbt suggests to use [selectors](https://docs.getdbt.com/reference/node-selection/yaml-selectors) for such cases, but it doesn't look like an easy fix to bring its support to dagster. Actually, dagster specifically [turns off](https://github.com/dagster-io/dagster/blob/bc81704666099720e8dacc927210301fc0542d22/python_modules/libraries/dagster-dbt/dagster_dbt/cloud/asset_defs.py#L519) existing selectors.

### How to reproduce?

Here is an example of difference in dbt behaviour:
image

### Deployment type

Dagster Cloud

### Deployment details

_No response_

### Additional information

_No response_

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

### PR DESCRIPTION
**Pull request by mudravrik**
## Summary & Motivation
Resolves #21593.

TLDR: Previously dagster-dbt added new --select arg to dbt Cloud command when subsetting some assets from the job. However this approach does not work since dbt 1.5.0 when multiple --select args have been supported. So dagster have to remove all selection args from cloud job command in order to run only subsetted models in dbt.

Also previously pure string operations were used to modify command, which can lead to not properly removed `--selector=xyz` args. So quasi-parser for dbt command was created to extract and remove all `--select` and `--selector` args from materialisation command.

I also slightly modified tests for subsetting by adding cases with options which should not be removed be dagster inside subsetting and cases for correct removing all `--select` flags from command. 

## How I Tested These Changes
Run unit test and a micro manual test with connecting to actual dbt Cloud project.

### CODE REVIEW

**Review by rexledesma — APPROVED** [review]
Great, thanks!
